network/p2p/p2pnode/internal/cache.go

```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+// disallowListFor is a helper function for disallowing a peer for a cause.
+// It adds the cause to the disallow list cache entity for the peerID and returns the updated list of causes for the peer.
+// Args:
+// - peerID: the peerID of the peer to be disallow-listed.
+// - cause: the cause for disallow-listing the peer.
+// Returns:
+// - the updated list of causes for the peer.
+// - error if the entity for the peerID is not found in the cache it returns ErrDisallowCacheEntityNotFound, which is a benign error.
+func (d *DisallowListCache) disallowListFor(peerID peer.ID, cause network.DisallowListedCause) ([]network.DisallowListedCause, error) {
+	adjustedEntity, adjusted := d.c.Adjust(makeId(peerID), func(entity flow.Entity) flow.Entity {
+		dEntity := mustBeDisallowListEntity(entity)
+		dEntity.causes[cause] = struct{}{}
+		return dEntity
+	})
+
+	if !adjusted {
+		// if the entity is not found in the cache, we return a benign error.
+		return nil, ErrDisallowCacheEntityNotFound
+	}
+
+	dEntity := mustBeDisallowListEntity(adjustedEntity)
+	updatedCauses := make([]network.DisallowListedCause, 0, len(dEntity.causes))
+	for cause := range dEntity.causes {
+		updatedCauses = append(updatedCauses, cause)
+	}
+
+	return updatedCauses, nil
+}
+
+// AllowFor removes a cause from the disallow list cache entity for the peerID.
+// Args:
+// - peerID: the peerID of the peer to be allow-listed.
+// - cause: the cause for allow-listing the peer.
+// Returns:
+// - the list of causes for which the peer is disallow-listed.
+// - error if the entity for the peerID is not found in the cache it returns ErrDisallowCacheEntityNotFound, which is a benign error.
+func (d *DisallowListCache) AllowFor(peerID peer.ID, cause network.DisallowListedCause) []network.DisallowListedCause {
+	adjustedEntity, adjusted := d.c.Adjust(makeId(peerID), func(entity flow.Entity) flow.Entity {
+		dEntity := mustBeDisallowListEntity(entity)
+		delete(dEntity.causes, cause)
+		return dEntity
+	})
+
+	if !adjusted {
+		// if the entity is not found in the cache, we return an empty list.
+		// we don't return a nil to be consistent with the case that entity is found but the list of causes is empty.
+		return make([]network.DisallowListedCause, 0)
+	}
+
+	dEntity := mustBeDisallowListEntity(adjustedEntity)
+	// returning a deep copy of causes (to avoid being mutated externally).
+	causes := make([]network.DisallowListedCause, 0, len(dEntity.causes))
+	for cause := range dEntity.causes {
+		causes = append(causes, cause)
+	}
+	return causes
+}
+
+// mustBeDisallowListEntity is a helper function for type assertion of the flow.Entity to disallowListCacheEntity.
+// It panics if the type assertion fails.
+// Args:
+// - entity: the flow.Entity to be type asserted.
+// Returns:
+// - the disallowListCacheEntity.
+func mustBeDisallowListEntity(entity flow.Entity) *disallowListCacheEntity {
```

Comment:
```suggestion
func (d *DisallowListCache) toDisallowListEntity(entity flow.Entity) *disallowListCacheEntity {
```

Reply:
The function does not require access to the shared state of the object `d`. Keeping it as a function (than a method) is more aligned with the [Principle of Least Privilege (PoLP)](https://en.wikipedia.org/wiki/Principle_of_least_privilege).